include/StochasticNoiseSimulator.hpp

```diff
@@ -3,189 +3,142 @@
-        Simulator(seed), qc(qc), step_number(step_number), step_fidelity(step_fidelity) {
-        dd->resize(qc->getNqubits());
+    StochasticNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc, const unsigned int stepNumber, const double stepFidelity, std::size_t seed):
+        Simulator<DDPackage>(seed), qc(qc), stepNumber(stepNumber), stepFidelity(stepFidelity) {
+        Simulator<DDPackage>::dd->resize(qc->getNqubits());
+        setMaxInstances();
     }
 
     StochasticNoiseSimulator(std::unique_ptr<qc::QuantumComputation>& qc,
-                             const std::string&                       noise_effects,
-                             double                                   noise_prob,
-                             long                                     stoch_runs,
-                             unsigned int                             step_number,
-                             double                                   step_fidelity,
-                             const std::string&                       recorded_properties):
-        qc(qc),
-        step_number(step_number), step_fidelity(step_fidelity) {
-        setNoiseEffects(noise_effects);
-        setRecordedProperties(recorded_properties);
-        initializeNoiseProbabilities(noise_prob);
-        setNumberOfRuns(stoch_runs);
-    }
+                             const std::string&                       cNoiseEffects,
+                             double                                   cNoiseProbability,
+                             std::optional<double>                    cAmpDampingProbability,
+                             double                                   cMultiQubitGateFactor,
+                             std::size_t                              cStochRuns,
+                             const std::string&                       recordedProperties,
+                             bool                                     unoptimizedSim,
+                             unsigned int                             stepNumber,
+                             double                                   stepFidelity,
+                             std::size_t                              seed = 0U):
+        Simulator<DDPackage>(seed),
+        qc(qc), stepNumber(stepNumber), stepFidelity(stepFidelity) {
+        // setNoiseEffects
+        for (const auto noise: cNoiseEffects) {
+            switch (noise) {
+                case 'A':
+                    noiseEffects.push_back(dd::amplitudeDamping);
+                    break;
+                case 'P':
+                    noiseEffects.push_back(dd::phaseFlip);
+                    break;
+                case 'D':
+                    noiseEffects.push_back(dd::depolarization);
+                    break;
+                default:
+                    throw std::runtime_error("Unknown noise operation '" + cNoiseEffects + "'\n");
+            }
+        }
```

Comment:
this does seem the same to me as in the deterministic simulator -> extract into a function and reuse.

Reply:
I moved the functionality into a static function of the StochasticSimulator which is reused in the deterministic simulator.